autoload/ale/code_action.vim

```diff
@@ -0,0 +1,73 @@
+function! ale#code_action#HandleCodeAction(code_action) abort
+    let l:current_buffer = bufnr('')
+    let l:existing_buffers = {}
+    let l:changes = a:code_action.changes
+
+    for l:file_code_edit in l:changes
+        let l:buf = ale#util#GetBufferNumber(l:file_code_edit.fileName)
+
+        if l:buf != -1
+            let l:existing_buffers[l:buf] = 1
+
+            if l:buf != l:current_buffer && ale#util#IsModified(l:buf)
+                call ale#util#Execute('echom ''Aborting action, file is unsaved''')
+                " Open buffer in question
+                call ale#util#Execute('buffer ' .  l:buf)
+
+                return
+            endif
+        endif
+    endfor
+
+    for l:file_code_edit in l:changes
+        let l:buf = ale#util#GetBufferNumber(l:file_code_edit.fileName)
+
+        if l:buf != -1
+            call ale#util#Execute('buffer ' . l:buf)
+        else
+            call ale#util#Execute('edit ' .  l:file_code_edit.fileName)
+            let l:buf = ale#util#GetBufferNumber(l:file_code_edit.fileName)
+        endif
+
+        let l:initial_pos = getpos('.')
+        let l:line_diff = 0
+
+        for l:code_edit in reverse(copy(l:file_code_edit.textChanges))
```

Comment:
Try what I was working on for applying changes to file instead: https://github.com/dense-analysis/ale/commit/ca2026bd13346cb7e1afeecb54dbec3da0bd5151

1. Search for an open buffer, and use `getbufline` to read it if the buffer exists. If the filename isn't open in a buffer, use `readfile()` to read the file.
2. Work out how to add in the text changes by slicing lines and then columns.
3. Write the list of lines back to the file again.

That should work pretty well. There may be some bug with it I hadn't spotted yet, but I had a series of tests against it that worked.

You can add that code in, and exclude the other code I had for context menus for now. I eventually plan on supporting pretty much all LSP code actions through either the right click context menu, or through a console menu. I might throw in a command for applying code actions.

Reply:
Thanks, this was super helpful! I fixed some indexing bugs that I encountered with my tests and implemented post-edit cursor movement.